test/e2e/storage/testsuites/snapshottable.go

```diff
@@ -179,25 +179,27 @@ func (s *snapshottableTestSuite) DefineTests(driver storageframework.TestDriver,
 			// In the following code by checking if the PV is not in the node.Status.VolumesInUse field we
 			// ensure that the volume is not used by the node anymore (an indicator that NodeUnstageVolume has
 			// already finished)
-			if framework.NodeOSDistroIs("windows") {
-				nodeName := pod.Spec.NodeName
-				gomega.Expect(nodeName).NotTo(gomega.BeEmpty(), "pod.Spec.NodeName must not be empty")
-
-				ginkgo.By(fmt.Sprintf("[init] waiting until the node=%s is not using the volume=%s", nodeName, pv.Name))
-				success := storageutils.WaitUntil(framework.Poll, f.Timeouts.PVDelete, func() bool {
-					node, err := cs.CoreV1().Nodes().Get(context.TODO(), nodeName, metav1.GetOptions{})
-					framework.ExpectNoError(err)
-					volumesInUse := node.Status.VolumesInUse
-					framework.Logf("current volumes in use: %+v", volumesInUse)
-					for i := 0; i < len(volumesInUse); i++ {
-						if strings.HasSuffix(string(volumesInUse[i]), pv.Name) {
-							return false
-						}
+			//
+			// We need to force this for non-Windows distros as well, as NodeUnpublishVolume only removes
+			// the bind mount, which doesn't force the caches to flush. It's possible to create empty
+			// snapshots if we don't wait (see #101279 for details).
+			nodeName := pod.Spec.NodeName
+			gomega.Expect(nodeName).NotTo(gomega.BeEmpty(), "pod.Spec.NodeName must not be empty")
+
+			ginkgo.By(fmt.Sprintf("[init] waiting until the node=%s is not using the volume=%s", nodeName, pv.Name))
```

Comment:
I think we should have 2 tests cases: One that takes a a snapshot online while it's still mounted, and one that takes a snapshot offline. 

I was hesitant when this workaround was initially added to Windows because I think the online snapshot case is the more common one, and the solution there is to have the application flush their data before taking a snapshot.

Reply:
but our "application" (`/bin/echo`) does not flush anything. With `/bin/sync` it works pretty well.

Reply:
Yah adding a sync to the tester pod so that we still have a test case for online snapshots makes sense to me. Then add another test for the offline snapshot. Doing a sync after the echo probably is a good idea in both cases to reduce flakiness.

Reply:
For the offline case, if we want to specifically verify that volume plugins are correctly flushing when unstaging, then it's better to not do the flush to mask plugin bugs.

But anyway, it seems like we've already turned these tests into the "offline" scenario for windows, so we can continue to do the same for linux, but we should find someone that has time to add the online scenario this release because I feel like the online scenario is the more common one.

Reply:
`sync` wasn't available in windows when I was working on https://github.com/kubernetes/kubernetes/pull/100377, FYI I opened https://github.com/kubernetes/kubernetes/issues/101172

Reply:
Feel free to correct me, but in an attempt to summarize the conversation and action items here:

1. We need to have two different test sets. With this PR (and https://github.com/kubernetes/kubernetes/pull/100057 ) the current tests have been converted to an "offline" scenario, where the container is no longer running at the time the snapshot is taken. This forces the volume to be unmounted and unpublished, which also results in a cache flush and disk write. If we assume this path forward (2 sets of tests), then we need to ensure all distros complete a NodeUnstageVolume call, and this PR is appropriate?
2. We need to create a series of online tests, where the container continues to run when we take the VolumeSnapshot. In this scenario we'll need to execute a `sync` to ensure that the cache is flushed and the data is persisted in the snapshot.

Is this understanding correct?

Reply:
Yup!

Reply:
So.. is this PR OK then?

Reply:
Opened up https://github.com/kubernetes/kubernetes/issues/101333 to track adding online shapshot tests

Reply:
I think the original test is considered more like offline test. The pod was deleted first and then snapshot is taken. The subtle thing here is that even when pod is deleted from api server, there might be a chance that volume is not finished unstaging yet.